### Pandas
Data scientists often work with large data sets that we need to inspect and manipulate. [Pandas](https://pandas.pydata.org/) is a popular package for handling datasets in Python, and will be often used in our data science courses. In this notebook we'll introduce the basics of loading and inspecting a data set with Pandas, which is typically imported with the name `pd`:

In [1]:
import pandas as pd

# File Inspection
Data sets are often saved as comma-separated values (CSV) files. You can inspect these with Excel or a simple text editor. Here we use the `read_csv` function from Pandas to load a dataset on the trees in Delft. This data set is made by the city of Delft and is [publicly available](https://data.overheid.nl/dataset/bomen-in-beheer-door-gemeente-delft).

In [2]:
trees_df = pd.read_csv('Bomen_in_beheer_door_gemeente_Delft.csv')

The dataset is loaded to the variable `trees_df` which is part of the DataFrame class. The DataFrame is the most used class of the Pandas package. Requesting `trees_df` gives a short overview of the dataset:

In [3]:
trees_df

X          Y     ID BORNAME  ELEMENTNUMMER  \
0      4.374816  52.029683  37732   bomen           1151   
1      4.373551  52.031048  37780   bomen           1184   
2      4.345117  51.994902   8154   bomen             43   
3      4.352882  52.013582  14766   bomen             89   
4      4.357415  51.993470  17389   bomen            252   
...         ...        ...    ...     ...            ...   
35378  4.343097  51.992807  35294   bomen            489   
35379  4.343493  51.993034  35296   bomen            491   
35380  4.343393  51.992935  35297   bomen            492   
35381  4.343578  51.992749  35301   bomen            495   
35382  4.343506  51.992816  35302   bomen            496   

                       BEHEERGROEP  AANLEGJAAR  \
0      Niet vrij uitgroeiende boom         NaN   
1      Niet vrij uitgroeiende boom         NaN   
2      Niet vrij uitgroeiende boom      1980.0   
3      Niet vrij uitgroeiende boom      1968.0   
4      Niet vrij uitgroeiende boom      1968.0   
...                            ...         ...   
35378        Boom Nader te bepalen      1983.0   
35379        Boom Nader te bepalen      1983.0   
35380        Boom Nader te bepalen      1983.0   
35381  Niet vrij uitgroeiende boom      1983.0   
35382  Niet vrij uitgroeiende boom      1983.0   

                            BOOMSORTIMENT BOOMSTATUS  \
0          Pyrus calleryana 'Chanticleer'        NaN   
1                          Acer campestre        NaN   
2             Ulmus x hollandica 'Vegeta'        NaN   
3      Aesculus hippocastanum 'Baumannii'        NaN   
4      Aesculus hippocastanum 'Baumannii'        NaN   
...                                   ...        ...   
35378                      Betula pendula        NaN   
35379                 Crataegus laevigata        NaN   
35380                      Betula pendula        NaN   
35381                  Fraxinus excelsior        NaN   
35382                      Acer campestre        NaN   

                              TAKVRIJE_ZONE  ... EXTRA_INFORMATIE_3  \
0                                   Bereikt  ...                IC2   
1      Onbereikbaar, snoeiingreep ongewenst  ...                IC2   
2      Onbereikbaar, snoeiingreep ongewenst  ...                IC3   
3      Onbereikbaar, snoeiingreep ongewenst  ...                IC1   
4      Onbereikbaar, snoeiingreep ongewenst  ...                IC3   
...                                     ...  ...                ...   
35378                                   NaN  ...                IC3   
35379                                   NaN  ...                IC3   
35380                               Bereikt  ...                IC3   
35381  Onbereikbaar, snoeiingreep ongewenst  ...                IC3   
35382  Onbereikbaar, snoeiingreep ongewenst  ...                IC3   

      GROENGEBIEDCODE                   GROENGEBIEDNAAM  GEMEENTE  \
0          01.16.0001  Bedrijventerrein Ypenburgsepoort     Delft   
1          01.16.0001  Bedrijventerrein Ypenburgsepoort     Delft   
2          01.25.0005                     Het Rode Dorp     Delft   
3          01.11.0002                      Centrum-west     Delft   
4          01.24.0006                  Voorhof-hoogbouw     Delft   
...               ...                               ...       ...   
35378      01.25.0003                       Gillisbuurt     Delft   
35379      01.25.0003                       Gillisbuurt     Delft   
35380      01.25.0003                       Gillisbuurt     Delft   
35381      01.25.0003                       Gillisbuurt     Delft   
35382      01.25.0003                       Gillisbuurt     Delft   

                                  BUURT             WIJK  \
0      Bedrijventerrein Ypenburgsepoort  16 Delftse Hout   
1      Bedrijventerrein Ypenburgsepoort  16 Delftse Hout   
2                         Het Rode Dorp     25 Buitenhof   
3                          Centrum-West    11 Binnenstad   
4                      Voorhof-Hoogbo

This DataFrame has 35384 rows that represent the different objects in the dataset and 28 columns that represent the different properties of each object. The output actually has 29 columns, but the first "column" doesn't have a label. This first column contains the index of each object, which in this case is a number, but can also be a string.  

Note that the DataFrame holds different variable types:
- `float` (X,Y)
- `int` (ID,ELEMENTNUMMER)
- `string` (BEHEERGROEP,BOOMSORTIMENT).

We don't have to inspect the table itself to get its size, the column labels, and the variable types. These are attributes of the DataFrame that can be accessed directly. For instance the size of the table is requested with attribute `shape`:

In [4]:
trees_df.shape # show the number of rows and columns

(35383, 28)

The column labels can be found with the attribute `columns`. Show the column labels below.

In [5]:
%%assignment
# YOUR CODE HERE
trees_df.columns

Index(['X', 'Y', 'ID', 'BORNAME', 'ELEMENTNUMMER', 'BEHEERGROEP', 'AANLEGJAAR',
       'BOOMSORTIMENT', 'BOOMSTATUS', 'TAKVRIJE_ZONE',
       'OMGEVINGSRISICOKLASSEN', 'AMBITIENIVEAU', 'HOOGTE', 'DIAMETER',
       'STANDPLAATS', 'BEHEERDER', 'EIGENAAR', 'EXTRA_INFORMATIE_2',
       'EXTRA_INFORMATIE_3', 'GROENGEBIEDCODE', 'GROENGEBIEDNAAM', 'GEMEENTE',
       'BUURT', 'WIJK', 'LIGGING', 'BEHEEROBJECTSOORT',
       'BEHEEROBJECTOMSCHRIJVING', 'OBJECTID'],
      dtype='object')

In [6]:
%%check
len(result) == 28
'BOOMSORTIMENT','TAKVRIJE_ZONE','GROENGEBIEDCODE' in result

Similarly, show the variable types with the attribute `dtypes`.

In [7]:
%%assignment
# YOUR CODE HERE
trees_df.dtypes

X                           float64
Y                           float64
ID                            int64
BORNAME                      object
ELEMENTNUMMER                 int64
BEHEERGROEP                  object
AANLEGJAAR                  float64
BOOMSORTIMENT                object
BOOMSTATUS                   object
TAKVRIJE_ZONE                object
OMGEVINGSRISICOKLASSEN       object
AMBITIENIVEAU                object
HOOGTE                       object
DIAMETER                    float64
STANDPLAATS                  object
BEHEERDER                    object
EIGENAAR                     object
EXTRA_INFORMATIE_2           object
EXTRA_INFORMATIE_3           object
GROENGEBIEDCODE              object
GROENGEBIEDNAAM              object
GEMEENTE                     object
BUURT                        object
WIJK                         object
LIGGING                      object
BEHEEROBJECTSOORT            object
BEHEEROBJECTOMSCHRIJVING     object
OBJECTID                    

In [ ]:
%%check
len(result) == 28
result['DIAMETER'].name == 'float64'
result['ID'].name == 'int64'
result['GROENGEBIEDCODE'].name == 'object'

The row labels are requested with the attribute `index`:

In [ ]:
trees_df.index

RangeIndex(start=0, stop=35383, step=1)

Currently, `index` returns an iterator that numbers each row. However, we can also specify to use one of the columns as row label. This is done by setting the option `index_col` of `read_csv` to the desired column label. 
In the current file, the column `ID` seems a good candidate as row label. Reload the dataset to the variable `df2` with the values of `ID` as row labels.

In [ ]:
%%assignment
# YOUR CODE HERE
df2 = pd.read_csv('Bomen_in_beheer_door_gemeente_Delft.csv', index_col='ID')
df2 # leave this to show the result

X          Y BORNAME  ELEMENTNUMMER  \
ID                                                  
37732  4.374816  52.029683   bomen           1151   
37780  4.373551  52.031048   bomen           1184   
8154   4.345117  51.994902   bomen             43   
14766  4.352882  52.013582   bomen             89   
17389  4.357415  51.993470   bomen            252   
...         ...        ...     ...            ...   
35294  4.343097  51.992807   bomen            489   
35296  4.343493  51.993034   bomen            491   
35297  4.343393  51.992935   bomen            492   
35301  4.343578  51.992749   bomen            495   
35302  4.343506  51.992816   bomen            496   

                       BEHEERGROEP  AANLEGJAAR  \
ID                                               
37732  Niet vrij uitgroeiende boom         NaN   
37780  Niet vrij uitgroeiende boom         NaN   
8154   Niet vrij uitgroeiende boom      1980.0   
14766  Niet vrij uitgroeiende boom      1968.0   
17389  Niet vrij uitgroeiende boom      1968.0   
...                            ...         ...   
35294        Boom Nader te bepalen      1983.0   
35296        Boom Nader te bepalen      1983.0   
35297        Boom Nader te bepalen      1983.0   
35301  Niet vrij uitgroeiende boom      1983.0   
35302  Niet vrij uitgroeiende boom      1983.0   

                            BOOMSORTIMENT BOOMSTATUS  \
ID                                                     
37732      Pyrus calleryana 'Chanticleer'        NaN   
37780                      Acer campestre        NaN   
8154          Ulmus x hollandica 'Vegeta'        NaN   
14766  Aesculus hippocastanum 'Baumannii'        NaN   
17389  Aesculus hippocastanum 'Baumannii'        NaN   
...                                   ...        ...   
35294                      Betula pendula        NaN   
35296                 Crataegus laevigata        NaN   
35297                      Betula pendula        NaN   
35301                  Fraxinus excelsior        NaN   
35302                      Acer campestre        NaN   

                              TAKVRIJE_ZONE OMGEVINGSRISICOKLASSEN  ...  \
ID                                                                  ...   
37732                               Bereikt               Verhoogd  ...   
37780  Onbereikbaar, snoeiingreep ongewenst               Algemeen  ...   
8154   Onbereikbaar, snoeiingreep ongewenst               Algemeen  ...   
14766  Onbereikbaar, snoeiingreep ongewenst               Algemeen  ...   
17389  Onbereikbaar, snoeiingreep ongewenst               Algemeen  ...   
...                                     ...                    ...  ...   
35294                                   NaN               Algemeen  ...   
35296                                   NaN               Algemeen  ...   
35297                               Bereikt               Algemeen  ...   
35301  Onbereikbaar, snoeiingreep ongewenst               Algemeen  ...   
35302  Onbereikbaar, snoeiingreep ongewenst               Algemeen  ...   

      EXTRA_INFORMATIE_3 GROENGEBIEDCODE                   GROENGEBIEDNAAM  \
ID                                                                           
37732                IC2      01.16.0001  Bedrijventerrein Ypenburgsepoort   
37780                IC2      01.16.0001  Bedrijventerrein Ypenburgsepoort   
8154                 IC3      01.25.0005                     Het Rode Dorp   
14766                IC1      01.11.0002                      Centrum-west   
17389                IC3      01.24.0006                  Voorhof-hoogbouw   
...                  ...             ...                               ...   
35294                IC3      01.25.0003                       Gillisbuurt   
35296                IC3      01.25.0003                       Gillisbuurt   
35297                IC3      01.25.0003                       Gillisbuurt   
35301                IC3      01.25.0003                       Gillisbuurt   
35302                IC3      01

In [ ]:
%%check 
hashresult == 1591492765


A quick overview of all the above information (column names, number of rows, variable types) can be made using the function `info`. It also provides the number of non-empty cells for each column, en the memory size. 

In [ ]:
trees_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35383 entries, 0 to 35382
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   X                         35383 non-null  float64
 1   Y                         35383 non-null  float64
 2   ID                        35383 non-null  int64  
 3   BORNAME                   35383 non-null  object 
 4   ELEMENTNUMMER             35383 non-null  int64  
 5   BEHEERGROEP               35383 non-null  object 
 6   AANLEGJAAR                27292 non-null  float64
 7   BOOMSORTIMENT             35383 non-null  object 
 8   BOOMSTATUS                1106 non-null   object 
 9   TAKVRIJE_ZONE             34041 non-null  object 
 10  OMGEVINGSRISICOKLASSEN    35383 non-null  object 
 11  AMBITIENIVEAU             35383 non-null  object 
 12  HOOGTE                    33592 non-null  object 
 13  DIAMETER                  1772 non-null   float64
 14  STANDP

The `DIAMETER` column only has 1772 rows with non-null values. The value of the other rows are indicated as `NaN`: Not a Number, which means these cells are empty. Missing values in data sets are common in practice, but we'll see that Pandas is very flexible with this. 

### Which statement(s) are correct?

In [ ]:
%%mmc file_inspection
OBJECTID has missing values
AANLEGJAAR has type int
Er zijn 28 kolommen

In [ ]:
%%check
hashresult == 3818810794